### PR TITLE
Postgres query plan support

### PR DESCRIPTION
**Pull request by rickcole**
Adds the following support for Postgres plans
1. Loading and rendering for EXPLAIN (FORMAT JSON)
2. Highlighting for magnitude differences in expected versus actual row counts
3. Crosslinks for common table expressions
4. Plan examples, primarily from "Learning to Optimize Queries"

### CODE REVIEW

File: query-graphs/src/postgres.ts [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            }
+        }
+
+        // Display the cardinality on the links between the nodes
+        let edgeLabel: string | undefined = undefined;
+        if (node.hasOwnProperty("Plan Rows") && typeof node["Plan Rows"] === "number") {
+            edgeLabel = formatMetric(node["Plan Rows"]);
+            if (node.hasOwnProperty("Actual Rows") && typeof node["Actual Rows"] === "number") {
+                //const num0 = Number(node["Plan Rows"]);
+                //const num1 = Number(node["Actual Rows"]);
```
Comment (vogelsgesang) [inline query-graphs/src/postgres.ts:119]: remove code which was commented out

File: query-graphs/src/postgres.ts [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+                //if (num0 > num1 * 10 || num0 * 10 < num1) {
```
Comment (vogelsgesang) [inline query-graphs/src/postgres.ts:120]: why prefer the "computeOrder" over this commented out condition?
Afaict, the `num0 > num1 * 10 || num0 * 10 < num1` would better correspond to the behavior expected by the user.

E.g. Plan Rows = 99 and Actual Rows = 100, would be highlighted by `order0 !== order1`, although being an estimate which is off-by-one, is perfectly fine...
`num0 > num1 * 10 || num0 * 10 < num1` would not highlight this case (as I would expect)

**Review by vogelsgesang — APPROVED** [review]

File: query-graphs/src/postgres.ts [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+// Function to generate nodes' display names based on their properties
+function generateDisplayNames(treeRoot: TreeNode) {
+    treeDescription.visitTreeNodes(
+        treeRoot,
+        function(node) {
+            switch (node.tag) {
+                case "Hash Join":
+                case "Nested Loop":
+                case "Merge Join":
+                    switch (node["Join Type"]) {
```
Comment (rickcole) [inline query-graphs/src/postgres.ts:188]: Join Type is actually in node.properties.

File: query-graphs/src/loader-utils.ts [hunk 4]
```diff
@@ -76,6 +76,16 @@ export function formatMetric(x: number): string {
     return x.toFixed(0) + sizes[idx];
 }
 
+// Compute order of magnitude
+export function computeOrder(x: number): number {
```
[the remark this answers is not in the document]
Comment (rickcole) [reply query-graphs/src/loader-utils.ts:80]: went back to simple 10x comparison

File: query-graphs/src/tree-rendering.ts [hunk 5]
```diff
@@ -864,7 +864,13 @@ export function drawQueryTree(target: HTMLElement, treeData: TreeDescription) {
         const linkLabelEnter = linkLabel
             .enter()
             .insert("text")
-            .classed("qg-link-label", true)
+            .attr("class", function(d) {
+                if (d.target.data.edgeLabel.includes(";")) {
```
[the remark this answers is not in the document]
Comment (rickcole) [reply query-graphs/src/tree-rendering.ts]: added edgeLabelClass

File: standalone-server/app.ts [hunk 6]
```diff
@@ -152,7 +154,8 @@ if (paramErrors.length) {
         if (fileFormat !== null) {
             loaders = [knownLoaders.get(fileFormat)];
         } else if (graphFile.endsWith(".json")) {
-            loaders = [loadHyperPlanFromText, loadJsonFromText];
+            // Try Postgres before Hyper so to differentiate between them
```
[the remark this answers is not in the document]
Comment (rickcole) [reply standalone-server/app.ts]: done

File: query-graphs/src/postgres.ts (hunk 1 above)
Comment (rickcole) [reply query-graphs/src/postgres.ts]: now using this logic

File: query-graphs/src/postgres.ts (hunk 2 above)
Comment (rickcole) [reply query-graphs/src/postgres.ts]: agreed

File: query-graphs/src/postgres.ts (hunk 3 above)
Comment (rickcole) [reply query-graphs/src/postgres.ts]: fixed

**Review by vogelsgesang — APPROVED** [review]

**Pull request merged**